src/hdmf/backends/hdf5/h5_utils.py

```diff
@@ -658,3 +658,15 @@ def valid(self):
         if isinstance(self.data, Dataset) and not self.data.id.valid:
             return False
         return super().valid
+
+    @property
+    def maxshape(self):
+        if 'maxshape' in self.io_settings:
+            return self.io_settings['maxshape']
+        elif hasattr(self.data, 'maxshape'):
+            return self.data.maxshape
+        elif hasattr(self, "shape"):
+            return self.shape
+        elif 'shape' in self.io_settings:
+            return self.io_settings['shape']
+        return None
```

Comment:
```suggestion
        else:
            return get_data_shape(self.data)
```

Did you have an example in mind for the `elif 'shape' in self.io_settings` condition? I think `shape` in `self.io_settings` will only be true if the `hasattr(self, "shape")` condition above is also true, but I might be missing a case.

I think there's also one more case where `maxshape` is not specified and the wrapped data object does not have a shape attribute (e.g. if `self.data` is a list/tuple), we could use `get_data_shape` in this case and it should return `None` if the shape cannot be inferred.

Reply:
Is there already an implementation of `get_data_shape` somewhere?

Reply:
Yes that's the one I was thinking of!

Reply:
Re `elif 'shape' in self.io_settings`... instantiating `H5DataIO(data=None, shape=(1,2))` will cause the base `DataIO` class to set `self.__data = None` and `self.__shape = shape`. I see now that if `self.__data is None`, then the `H5DataIO` constructor will assign `self.__iosettings['shape'] = self.shape`.

It's a little bit weird to have this redundant information stored in 2 locations in the same class. I guess I was hedging against the possibility of them getting out of sync somehow. I don't mind removing this condition.